### PR TITLE
Generic settings change prototype

### PR DESCRIPTION
**Pull request by LordFarin**
While I'm not completely happy with everything, it is now in a functioning state. I think it's good to work from here. This pull request addresses #90.

### CODE REVIEW

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java [hunk 1]
```diff
@@ -198,6 +193,11 @@ public void run() {
 
     }
     
+    @SettingListener( setting = ApplicationSettings.Setting.AUTOCOMPLETE_ENABLED )
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: It's debatable whether we want to use the same annotation for a setting _listener_ and a setting _changer_. Anyhow the approach should be symmetrical so maybe a SettingWriter annotation is in order.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 2]
```diff
@@ -107,12 +117,76 @@ public boolean getLineWrapStatus(){
...
-        setProperty(AUTOCOMPLETE_STATUS, String.valueOf(autoCompleteStatus));
+    public void registerSettingListeners(Object o) {
+        for(Method m : o.getClass().getMethods()) {
+            for(Annotation a : m.getAnnotations()) {
+                if( a instanceof SettingListener ) {
+                    Setting s = ((SettingListener) a).setting();
+                    //TODO 20160323 Type validation
+                    settingListeners.get(s).add(Pair.of(o,m));
+                    try {
+                        m.invoke(o, getSettingValue(s));
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: This line provides convenient enhancement of the registration with a first "update" of the setting with the current value.
Also, although I Javadoc'ed the annotation now, methods not fulfilling the contract should be ignored here.

**Comment by sebbrudzinski** [comment]
Thanks @LordFarin I'll try to review this weekend.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java [hunk 3]
```diff
@@ -82,21 +83,26 @@ public void propertyChange(PropertyChangeEvent evt) {
                     autoCompletion.setAutoActivationDelay((Integer) evt.getNewValue());
                 } else if (evt.getPropertyName().equals(ApplicationSettings.LINEWRAP_STATUS) && rSyntaxTextArea != null) {
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: What about remaining properties?

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 4]
```diff
@@ -26,13 +32,17 @@
     private static final String LATEX_PATH      = "latex.path";
     public static final String AUTOCOMPLETE_DELAY = "autocomplete.delay";
     public static final String LINEWRAP_STATUS = "linewrap.status";
-    public static final String AUTOCOMPLETE_STATUS = "autocomplete.status";
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: Why was this removed?

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 5]
```diff
@@ -107,12 +117,76 @@ public boolean getLineWrapStatus(){
         return Boolean.parseBoolean(getProperty(LINEWRAP_STATUS));
     }
     
-    public void setAutoCompleteStatus(boolean autoCompleteStatus){
-        pcs.firePropertyChange(AUTOCOMPLETE_STATUS, getAutoCompleteStatus(), autoCompleteStatus);
-        setProperty(AUTOCOMPLETE_STATUS, String.valueOf(autoCompleteStatus));
+    public void registerSettingListeners(Object o) {
+        for(Method m : o.getClass().getMethods()) {
+            for(Annotation a : m.getAnnotations()) {
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: Method exposes "getAnnotation(Class<T>)", which allows to specify the annotation we wish to retrieve. This would simplify a lot in this method.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 6]
```diff
@@ -107,12 +117,76 @@ public boolean getLineWrapStatus(){
...
+                    //TODO 20160323 Type validation
+                    settingListeners.get(s).add(Pair.of(o,m));
+                    try {
+                        m.invoke(o, getSettingValue(s));
+                    } catch (IllegalAccessException ex) {
+                        Exceptions.printStackTrace(ex);
+                    } catch (IllegalArgumentException ex) {
+                        Exceptions.printStackTrace(ex);
+                    } catch (InvocationTargetException ex) {
+                        Exceptions.printStackTrace(ex);
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: We can have a multi-catch here

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java [hunk 7]
```diff
@@ -107,12 +117,76 @@ public boolean getLineWrapStatus(){
...
+    private void fireSettingChange(Setting setting, String value) {
+        for( Pair<Object,Method> p : settingListeners.get(setting)) {
+            try {
+                p.second().invoke(p.first(), value);
+            } catch (IllegalAccessException ex) {
+                Exceptions.printStackTrace(ex);
+            } catch (IllegalArgumentException ex) {
+                Exceptions.printStackTrace(ex);
+            } catch (InvocationTargetException ex) {
+                Exceptions.printStackTrace(ex);
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: We can have a multi-catch here

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/SettingListener.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/SettingListener.java]: We must use project's license template in all files

**Comment by sebbrudzinski** [comment]
I've left some comments/questions/improvements suggestions.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java (hunk 3 above)
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: I wanted to bring the mechanism to the master branch ASAP in order for everyone to use it. Migration of the other settings is up next.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java (hunk 4 above)
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/settings/ApplicationSettings.java]: This was replaced by the new mechanism, which defines the string in the ApplicationSettings.Settings enum.

File: OpenLaTeXStudio/EditorModule/nbproject/project.properties [hunk 9]
```diff
@@ -1,2 +1,2 @@
-javac.source=1.6
+javac.source=1.7
```
Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/nbproject/project.properties]: Multi-catch requires 1.7.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java (hunk 1 above)
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: This made me wonder.. if I wanted to fire the same property change event from two different classes, wouldn't I fall into infinite loop, since we have the same annotations for writers and listeners?

Comment (LordFarin) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/LaTeXSettingsPanel.java]: Not unless the ApplicationSettings.setSettingValue is actually called from the SettingListener annotated method, I think (which here is in store()). And it would seem obvious to me that such a thing is a very bad idea.
I still dislike the disparity between the annotation and the setSettingValue method, though.

File: OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java (hunk 3 above)
Comment (sebbrudzinski) [inline OpenLaTeXStudio/EditorModule/src/latexstudio/editor/EditorTopComponent.java]: To be honest, I'd feel more confident merging those changes, once all of the settings are migrated to the new system. At this point it's not quite easy for me to figure out how the final result will look like. I can even hold off all other tickets adding/editing properties for the sake of finishing this first (we do not have many of them anyways).

**Comment by LordFarin** [comment]
All settings have been moved. There has been a small paradigm shift in ApplicationSettings.Setting handling which simplifies setting change calls and can be considered the first move towards a SettingWriter annotation.
